### PR TITLE
feat(workflows): implement basic CI pipeline

### LINKED ISSUE #4: Add CI
It's always a good idea to enforce some code quality in your master branch. Even this early on.

Thankfully this is quite straightforward in the rust ecosystem.

Implement the following pipeline
- [x] Lint (rustfmt + clippy + commitlint)
- [x] Test (cargo)
- [x] Release (semver)
- [x] Publish (always dry run)

Crates.io [doesn't need you to build binaries](https://rust-cli.github.io/book/tutorial/packaging.html) (although there are good reasons to build them anyway). None of those good reasons are pressing enough for this project (yet).

### PR DESCRIPTION
**Pull request by CameronMatthew**
Closes #4.

Add...
- Rust lint (clippy + rustfmt)
- Commit lint
- Rust test
- Semantic release
- Publish (dry run).

### CODE REVIEW

**Comment by CameronMatthew** [comment]
I have made the above checks "required" for merge into main.

**Comment by CameronMatthew** [comment]
Commit format is as specified in [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      - name: Install commitlint
+        run: npm install --save-dev @commitlint/{cli,config-conventional}
+
+      - name: Commitlint (PR)
+        if: github.event_name == 'pull_request'
+        run: npx commitlint --from ${{ github.event.pull_request.base.sha }}
+    
+      - name: Commitlint (Push)
+        if: github.event_name == 'push'
+        run: npx commitlint --from ${{ github.event.before }}
```
Comment (siliconlad) [inline .github/workflows/lint.yml:50]: Based on my limited reading, sometimes `github.event.before` is empty? Like the first commit of a branch?

**Review by siliconlad — COMMENTED** [review]
I propose we omit release and publish from the pipeline until we have something reasonably stable in place. Perhaps the same with `commitlint` too.

Otherwise every second commit is going to be 'BREAKING CHANGE: redid the whole goddamn thing.'. And `feotensor` is going to be on v201.67.104 before we know it.

Like it otherwise though!

File: .github/workflows/lint.yml (hunk 1 above)
Comment (CameronMatthew) [reply .github/workflows/lint.yml]: I also think that releases and versioning are essential as soon as someone depends on this package. Already I could start using what you've written to get started on other projects. So might as well start releasing. The version number is not for marketing or for aesthetics - so there's no harm on being on v201.0.0 as far as I'm concerned!

TL;DR releases are useful immediately so we might as well start making them.

Comment (CameronMatthew) [reply .github/workflows/lint.yml]: I also think you'll be surprised at how little your API will change. The interface you created won't change that much with the refactoring we have planned. Introducing new stuff is not a breaking change. So I think it'll be fine.

**Comment by CameronMatthew** [comment]
Merged your fixup commit into mine so this should pass commitlint.

File: .github/workflows/lint.yml (hunk 1 above)
Comment (siliconlad) [reply .github/workflows/lint.yml]: Hmm, yeah fair enough. If you can already start using this for other projects might as well release versions so we can keep track of things.

I will read more into this commitlint to understand how it works.

**Review by siliconlad — DISMISSED** [review]
Let's give this a go!

**Comment by siliconlad** [comment]
Just read up (maybe re-read?) on Conventional Commits and have a couple of questions:
1. What `type`s are we using? `config-conventional` introduces a couple others:
    ```
    build: Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
    ci: Changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs)
    docs: Documentation only changes
    feat: A new feature
    fix: A bug fix
    perf: A code change that improves performance
    refactor: A code change that neither fixes a bug nor adds a feature
    style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
    test: Adding missing tests or correcting existing tests
    ```
    Based on this, surely we should be using something like the `ci` type for the commit instead of `feat`.
2. Can we not still adhere to `comver` if we remap `fix` and others to `minor` instead of `patch`? This way our commits can still display `fix` and `feat` which might be useful in a git history but have `comver` which is useful for when others use your package.

**Comment by siliconlad** [comment]
Also, do we want to release a new version on every merge to `main` (I believe that this is the current behaviour)? I feel like the release and publish should be gated so that we can pool change together before a release is created.

**Comment by CameronMatthew** [comment]
> Just read up (maybe re-read?) on Conventional Commits and have a couple of questions:
> 
> 1. What `type`s are we using? `config-conventional` introduces a couple others:
>    ```
>    build: Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
>    ci: Changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs)
>    docs: Documentation only changes
>    feat: A new feature
>    fix: A bug fix
>    perf: A code change that improves performance
>    refactor: A code change that neither fixes a bug nor adds a feature
>    style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
>    test: Adding missing tests or correcting existing tests
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    Based on this, surely we should be using something like the `ci` type for the commit instead of `feat`.
> 2. Can we not still adhere to `comver` if we remap `fix` and others to `minor` instead of `patch`? This way our commits can still display `fix` and `feat` which might be useful in a git history but have `comver` which is useful for when others use your package.
> 
> Perhaps a basic CONTRIBUTING.md is in order to outline this?

Yeah I'd be for remapping to comver. That's a good shout. I'll make that change. Yeah I should have used the ci type lol. Will fix that too.

**Comment by CameronMatthew** [comment]
> Also, do we want to release a new version on every merge to `main` (I believe that this is the current behaviour)? I feel like the release and publish should be gated so that we can pool changes together before a release is created.

For security and clarity reasons - the mantra of semantic versioning is that every commit is a new release. Even the smallest change such as bumping a dependency version is worth releasing to users immediately - as it could be a security upgrade. I see no reason for us to pool releases at the moment.

**Comment by CameronMatthew** [comment]
Done @siliconlad.

**Comment by CameronMatthew** [comment]
> Also, do we want to release a new version on every merge to `main` (I believe that this is the current behaviour)? I feel like the release and publish should be gated so that we can pool changes together before a release is created.

Also with the release rules specification I defined here, you can merge to main without committing if your changes fall under "refactor", "ci", "style", "docs" or "test".

File: Cargo.toml [hunk 2]
```diff
@@ -2,5 +2,10 @@
 name = "feotensor"
 version = "0.1.0"
 edition = "2021"
+description = "A Scientific Rust library for tensor operations"
```
Comment (siliconlad) [inline Cargo.toml:5]:
```suggestion
description = "A tensor library for scientific computing in Rust."
```

**Review by siliconlad — APPROVED** [review]

**Review by siliconlad — APPROVED** [review]

**Comment by CameronMatthew** [comment]
Oh man, neither of us can approve this now lol. Let me rebase so you can approve

**Comment by CameronMatthew** [comment]
A ffs, just going to bypass protection rules. LOL
